pkg/apiserver/metrics/deployments.go

```diff
@@ -0,0 +1,168 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"k8s.io/kubernetes/pkg/api/rest"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"golang.org/x/net/context"
+)
+
+var (
+	variableLabels = []string{"name", "namespace"}
+
+	deploymentStatusObservedGenerationDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_observed_generation",
+		"The observed generation of this deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_replicas",
+		"The number of replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusUpdatedReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_updated_replicas",
+		"The number of updated replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusAvailableReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_available_replicas",
+		"The number of available replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusUnavailableReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_unavailable_replicas",
+		"The number of unavailable replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+)
+
+// DeploymentCollector collects various metrics of deployment objects.
+type DeploymentCollector struct {
+	Storage rest.StandardStorage
+}
+
+// Returns a new instance of DeploymentCollector.
+func NewDeploymentsCollector(s rest.StandardStorage) *DeploymentCollector {
+	return &DeploymentCollector{
+		Storage: s,
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *DeploymentCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- deploymentStatusObservedGenerationDesc
+	ch <- deploymentStatusReplicasDesc
+	ch <- deploymentStatusUpdatedReplicasDesc
+	ch <- deploymentStatusAvailableReplicasDesc
+	ch <- deploymentStatusUnavailableReplicasDesc
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *DeploymentCollector) Collect(ch chan<- prometheus.Metric) {
+	var err error
+	deployments, err := c.Storage.List(context.Background(), nil)
+	if err != nil {
+		glog.Errorf("Error while collecting deployment metrics: %s", err.Error())
+	}
+
+	for _, d := range deployments.(*extensions.DeploymentList).Items {
+		err = addMetrics(ch, d)
+		if err != nil {
+			glog.Errorf("Error while collecting deployment metrics: %s", err.Error())
+		}
+	}
+}
+
+// addMetrics adds Metrics for a single deployment to the specified channel
+func addMetrics(ch chan<- prometheus.Metric, d extensions.Deployment) error {
```

Comment:
There are a couple of reasons I started that question - the apiservers
at the heart of Kubernetes are expected to be dumb relays of state
(closer than an object store than a classic API).  Requiring new and
exciting metrics to be collected specific to each object in some ways
(not all) cuts against that abstraction.  I think we'd probably say we
want to let other components calculate many of the metrics described
here for other reasons, even if we ultimately decide to provide
metrics here.

The controllers, on the other hand, *do* see all changes, have
optimitized data structures specifically for accepting those changes,
and make domain specific decisions.  So even adding info on the
apiserver may not provide *enough* metrics - certainly not those that
relate to business logic decisions (like perm failed deployments, for
instance).  Since there are also the actors, their clock (their rate)
will in practice be as accurate as the apiservers.  We would
potentially have to deal with double metrics. But the apiserver may as
well (and we can high-water-mark metrics on the controllers).

Part of the point of a proposal is to flush these issues out and help
you guys make the right technical decision, as well as capturing the
resulting discussion in a concrete form for others.